site/src/paradox/internals/Coders.md

```diff
@@ -25,7 +26,56 @@ Beam provides built-in Coders for various basic Java types (`Integer`, `Long`, `
 import com.spotify.scio.values.SCollection
 
 case class Foo(x: Int, s: String)
-def sc: SCollection[Foo] = ??? // Beam will need an org.apache.beam.sdk.coders.Coder[Foo]
+def coll: SCollection[Foo] = ??? // Beam will need an org.apache.beam.sdk.coders.Coder[Foo]
+```
+
+### When are `Coder` used ?
+
+#### Shuffling data
+
+Whenever intermediate data is shuffled, Beam will need to serialize and deserialize that data to transfer it between workers. In Scio any `*byKey` (`groupByKey`, `reduceByKey`, etc.) transform will trigger a shuffle.
+
+#### Cluster scaling up and down
+
+When the runner scales up and down your cluster size, data needs to be redistributed between workers. Beam therefore needs to transfer data over the network, which means serializing and deserializing it by using `Coder`.
+
+#### GroupByKey
+
+Grouping by key uses `Coder` for two reasons: First, as we have already seen, GBK triggers a shuffle, and therefore go through a serialization / deserialization cycle.
+
+Second, grouping elements by key means that Beam needs to compare them and be able to decide whether two instances are equal. In Beam, and in the context of a `groupByKey` (or any `*byKey` operation), the equality of keys is tested by comparing their serialized form.
+
+Let's say we have defined a class `Identifier` and we use it as the key in a `groupByKey` transform:
+
+```scala
+import com.spotify.scio.values.SCollection
+
+val coll: SCollection[(Identifier, Foo)] = ???
+val grouped: SCollection[(Identifier, Iterable[Foo])] = coll.groupByKey()
+```
+
+To decide whether two instances of `Identifier` `id1` and `id2` are equal, Beam will compare them after serialization. For example:
+
+```scala
+// (pseudo-code)
+coder.encode(id1) // 00010011
+coder.encode(id2) // 00010010
+// -> i1 and i2 are NOT equal
+```
+
+**When they are used to test equality, coders are required to be deterministic.**
+If a non-deterministic `Coder` is used to test equality, an exception is thrown:
+
+```scala mdoc:reset:invisible
+import com.spotify.scio.ScioContext
+val sc = ScioContext.forTest()
```

Comment:
should we name this more explicit? perhaps `context` or `scioContext`?

Reply:
We call it `sc` pretty much everywhere in docs and example. I agree it's no ideal but consistency is nice ^^